hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestPayloadDeprecationFlow.scala

```diff
@@ -168,6 +184,162 @@ class TestPayloadDeprecationFlow extends SparkClientFunctionalTestHarness {
+      (12, 1L, "rider-X", "driver-X", 20.10, "D", "12.1", 12, 1),
+      (11, 2L, "rider-Y", "driver-Y", 27.70, "u", "11.1", 11, 1))
+    val firstUpdate = spark.createDataFrame(firstUpdateData).toDF(columns: _*)
+    firstUpdate.write.format("hudi").
+      option(OPERATION.key(), "upsert").
+      option(HoodieCompactionConfig.INLINE_COMPACT.key(), "false").
+      mode(SaveMode.Append).
+      save(basePath)
+    // Validate table version.
+    metaClient = HoodieTableMetaClient.reload(metaClient)
+    assertEquals(9, metaClient.getTableConfig.getTableVersion.versionCode())
+    val firstUpdateInstantTime = metaClient.getActiveTimeline.getInstants.get(1).requestedTime()
+
+
+    // 3. Add an update. This is expected to trigger the upgrade
+    val compactionEnabled = if (tableType.equals(HoodieTableType.MERGE_ON_READ.name())) {
+      "true"
+    } else {
+      "false"
+    }
+    val secondUpdateData = Seq(
+      (12, 3L, "rider-CC", "driver-CC", 33.90, "i", "12.1", 12, 1),
+      (9, 4L, "rider-DD", "driver-DD", 34.15, "i", "9.1", 9, 1),
+      (12, 5L, "rider-EE", "driver-EE", 17.85, "i", "12.1", 12, 1))
+    val secondUpdate = spark.createDataFrame(secondUpdateData).toDF(columns: _*)
+    secondUpdate.write.format("hudi").
+      option(OPERATION.key(), "upsert").
+      option(HoodieCompactionConfig.INLINE_COMPACT.key(), compactionEnabled).
+      option(HoodieCompactionConfig.INLINE_COMPACT_NUM_DELTA_COMMITS.key(), "1").
+      mode(SaveMode.Append).
+      save(basePath)
+    // Validate table version as 9.
+    metaClient = HoodieTableMetaClient.reload(metaClient)
+    assertEquals(9, metaClient.getTableConfig.getTableVersion.versionCode())
+    assertEquals(payloadClazz, metaClient.getTableConfig.getLegacyPayloadClass)
+    val compactionInstants = metaClient.getActiveTimeline.getCommitsAndCompactionTimeline.getInstants
+    val foundCompaction = compactionInstants.stream().anyMatch(i => i.getAction.equals("commit"))
+    assertTrue(foundCompaction)
+
+    // 4. Add a trivial update to trigger payload class mismatch.
+    val thirdUpdateData = Seq(
+      (12, 3L, "rider-CC", "driver-CC", 33.90, "i", "12.1", 12, 1))
+    val thirdUpdate = spark.createDataFrame(thirdUpdateData).toDF(columns: _*)
+    if (!payloadClazz.equals(classOf[MySqlDebeziumAvroPayload].getName)) {
+      assertThrows[HoodieException] {
+        thirdUpdate.write.format("hudi").
+          option(OPERATION.key(), "upsert").
+          option(HoodieCompactionConfig.INLINE_COMPACT.key(), "false").
+          option(HoodieCompactionConfig.INLINE_COMPACT_NUM_DELTA_COMMITS.key(), "1").
+          option(HoodieTableConfig.PAYLOAD_CLASS_NAME.key(),
+            classOf[MySqlDebeziumAvroPayload].getName).
+          mode(SaveMode.Append).
+          save(basePath)
+      }
+    }
+
+    // 5. Add a delete.
+    val fourthUpdateData = Seq(
+      (12, 3L, "rider-CC", "driver-CC", 33.90, "i", "12.1", 12, 1),
```

Comment:
We should have some test that validates that the delete by the op field is working as expected.